### PR TITLE
Linear layer

### PR DESCRIPTION
**Pull request by jjsjann123**
To support fusion with linear layer, we did:
1. new operator `add_optional` that supports add with optional[Tensor];
2. decompose pass that breaks `linear` layer into `matmul` and `add_optional`;
3. parser rule added for `add_optional` and `linear`;
4. python test;

### CODE REVIEW

**Comment by jjsjann123** [comment]
Noticed that ToT code base is broken on my local build. After reverting bd7ae53adf76f3851a4a539b1ff4062544d7ca7c I'm passing all tests.

**Comment by csarofeen** [comment]
cc @naoyam for that commit you're having issues with.

File: torch/csrc/jit/codegen/cuda/graph_fuser.cpp [hunk 1]
```diff
@@ -1310,12 +1310,58 @@ void traverseProfileIValues(
...
+  auto graph = block->owningGraph();
+  for (Node* n : linear_nodes) {
+    WithInsertPoint guard(n);
+    auto weight_t = graph->insertNode(graph->create(aten::t, {n->input(1)}, 1));
+    auto matmul = graph->insertNode(
+        graph->create(aten::matmul, {n->input(0), weight_t->output()}, 1));
+    auto input_tensor_type = n->input(0)->type()->cast<c10::TensorType>();
+    auto mat0_size = input_tensor_type->sizes().concrete_sizes();
+    auto mat1_size =
+        n->input(1)->type()->cast<c10::TensorType>()->sizes().concrete_sizes();
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/graph_fuser.cpp:1339]: Are these actually concrete sizes? Or this is this related to inserting profiling nodes?

Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/graph_fuser.cpp:1339]: This is propagating profiled sizes. So they are concrete but it shouldn't break our dynamic size support

**Comment by naoyam** [comment]
> Noticed that ToT code base is broken on my local build. After reverting [bd7ae53](https://github.com/csarofeen/pytorch/commit/bd7ae53adf76f3851a4a539b1ff4062544d7ca7c) I'm passing all tests.

Can you show the error output? I don't see any test error with my build.

**Comment by jjsjann123** [comment]
> > Noticed that ToT code base is broken on my local build. After reverting [bd7ae53](https://github.com/csarofeen/pytorch/commit/bd7ae53adf76f3851a4a539b1ff4062544d7ca7c) I'm passing all tests.
> 
> Can you show the error output? I don't see any test error with my build.

I'm getting segfaults from both cpp test as well as python test (with release build, haven't tried debug build yet).
this is the broken cpp test: ` ../build/bin/test_jit --gtest_filter="NVFuserTest.FusionReductionSchedulerNoODimShmoo_CUDA"`

**Comment by naoyam** [comment]
> > > Noticed that ToT code base is broken on my local build. After reverting [bd7ae53](https://github.com/csarofeen/pytorch/commit/bd7ae53adf76f3851a4a539b1ff4062544d7ca7c) I'm passing all tests.
> > 
> > 
> > Can you show the error output? I don't see any test error with my build.
> 
> I'm getting segfaults from both cpp test as well as python test (with release build, haven't tried debug build yet).
> this is the broken cpp test: ` ../build/bin/test_jit --gtest_filter="NVFuserTest.FusionReductionSchedulerNoODimShmoo_CUDA"`

Thanks. I'll check it with release build.

**Comment by naoyam** [comment]
> > > > Noticed that ToT code base is broken on my local build. After reverting [bd7ae53](https://github.com/csarofeen/pytorch/commit/bd7ae53adf76f3851a4a539b1ff4062544d7ca7c) I'm passing all tests.
> > > 
> > > 
> > > Can you show the error output? I don't see any test error with my build.
> > 
> > 
> > I'm getting segfaults from both cpp test as well as python test (with release build, haven't tried debug build yet).
> > this is the broken cpp test: ` ../build/bin/test_jit --gtest_filter="NVFuserTest.FusionReductionSchedulerNoODimShmoo_CUDA"`
> 
> Thanks. I'll check it with release build.

That test seems to work fine with my build, but I see several other test failures. For example:

```
[ RUN      ] NVFuserTest.FusionBiasGeluFwd_CUDA
unknown file: Failure
C++ exception with description "The size of tensor a (35832268) must match the size of tensor b (0) at non-singleton dimension 2
Exception raised from infer_size at ../aten/src/ATen/ExpandUtils.cpp:24 (most recent call first):
```

I tried older commits to see when this failure started to show up but couldn't figure it out yet. Am I the only one seeing this error?

**Review by csarofeen — APPROVED** [review]
